### PR TITLE
feat: Add Table of Contents component

### PR DESCRIPTION
**Pull request by dlnr**
Styled after the design for amsterdam.nl, Bas hinted that the chevrons were floating so I removed them. 

This component has no added scripts to set an active state while scrolling or folding children (so chevron also not needed). 

We need to await futher refinement for this component, this is just the basics.

### CODE REVIEW

File: packages/css/src/components/table-of-contents/table-of-contents.scss [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+@mixin reset {
+  box-sizing: border-box;
+  margin-block: 0;
+  padding-inline: 0;
+}
+
+.ams-table-of-contents {
+  display: flex;
+  flex-direction: column;
+  font-family: var(--ams-table-of-contents-font-family, inherit);
```
Comment (alimpens) [inline packages/css/src/components/table-of-contents/table-of-contents.scss:17]: What does the fallback do?

File: proprietary/tokens/src/components/ams/table-of-contents.tokens.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
+{
+  "ams": {
+    "table-of-contents": {
+      "color": { "value": "{ams.color.primary-black}" },
```
Comment (alimpens) [inline proprietary/tokens/src/components/ams/table-of-contents.tokens.json:4]: Can you check if all tokens are used in the CSS? At a quick glance, they aren't

File: packages/css/src/components/table-of-contents/README.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<!-- @license CC0-1.0 -->
+
+# Table of Contents
+
+A table of contents is a list of links to sections of a document.
```
Comment (alimpens) [inline packages/css/src/components/table-of-contents/README.md:5]:
```suggestion
A list of links corresponding to the content sections on the page.
It helps users to easily navigate to different sections on the same page.
```

File: packages/react/src/TableOfContents/TableOfContents.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+export type TableOfContentsProps = {
+  title?: string
+} & PropsWithChildren<HTMLAttributes<HTMLDivElement>>
+
+const TableOfContentsRoot = forwardRef(
+  ({ children, className, title, ...restProps }: TableOfContentsProps, ref: ForwardedRef<HTMLDivElement>) => {
+    return (
+      <nav {...restProps} ref={ref} className={clsx('ams-table-of-contents', className)}>
+        {title && <div className="ams-table-of-contents__title">{title}</div>}
```
[the remark this answers is not in the document]
Comment (dlnr) [reply packages/react/src/TableOfContents/TableOfContents.tsx:19]: Right also in my other PR....

File: storybook/config/storybook-overrides.css [hunk 5]
```diff
@@ -1,5 +1,19 @@
...
+    &:nth-child(2) {
+      padding-block-end: 50vh;
+    }
+  }
+  #grid-row-test-toc {
+    position: sticky;
+    top: 24px;
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (dlnr) [reply storybook/config/storybook-overrides.css:16]: Forgot to remove this, I had an article with functioning scroll observer.

File: packages/react/src/TableOfContents/TableOfContents.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        <ol className="ams-table-of-contents__list">{children}</ol>
```
[the remark this answers is not in the document]
Comment (VincentSmedinga) [reply packages/react/src/TableOfContents/TableOfContents.tsx:20]: Sections on a page have a well-considered order, which makes `ol` a good option.

A site is more a collection than an ordered set of pages. So when using this component as a site menu, a `ul` would make more sense.

I don’t think we want to use or allow `ol` attributes like `type`, `start` and `reversed`. Even an unordered list has an order of itself. So maybe a `ul` is the safest option?

File: packages/react/src/TableOfContents/TableOfContentsLink.tsx [hunk 7]
```diff
@@ -0,0 +1,40 @@
+/**
+ * @license EUPL-1.2+
+ * Copyright Gemeente Amsterdam
+ */
+
+import { clsx } from 'clsx'
+import { forwardRef } from 'react'
+import type { AnchorHTMLAttributes, ForwardedRef } from 'react'
+
+type BackgroundName = 'default' | 'light' | 'dark'
```
[the remark this answers is not in the document]
Comment (VincentSmedinga) [reply packages/react/src/TableOfContents/TableOfContentsLink.tsx:10]: Discussed this morning, we don’t want it on a dark background.

File: proprietary/tokens/src/components/ams/table-of-contents.tokens.json [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      "font-family": { "value": "{ams.text.font-family}" },
+      "font-size": { "value": "{ams.text.level.5.font-size}" },
+      "font-weight": { "value": "{ams.text.font-weight.normal}" },
+      "gap": { "value": "{ams.space.inside.lg}" },
```
Comment (VincentSmedinga) [inline proprietary/tokens/src/components/ams/table-of-contents.tokens.json:8]: The gap is 16 pixels in Figma.

```suggestion
      "gap": { "value": "{ams.space.inside.md}" },
```

File: proprietary/tokens/src/components/ams/table-of-contents.tokens.json [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        },
+        "on-background-light": {
+          "color": { "value": "{ams.link-appearance.on-background-light.color}" },
+          "hover": {
+            "color": { "value": "{ams.link-appearance.on-background-light.color}" }
+          }
+        }
+      },
+      "list": {
+        "gap": { "value": "{ams.space.inside.lg}" },
```
Comment (VincentSmedinga) [inline proprietary/tokens/src/components/ams/table-of-contents.tokens.json:34]: This one should be 16 as well.

```suggestion
        "gap": { "value": "{ams.space.inside.md}" },
```

File: packages/css/src/components/table-of-contents/table-of-contents.scss [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  gap: var(--ams-table-of-contents-list-gap);
+  list-style: none;
+
+  .ams-table-of-contents__list {
+    padding-block-start: var(--ams-table-of-contents-list-gap);
+    padding-inline-start: var(--ams-table-of-contents-list-list-padding-inline-start);
+  }
+
+  @include text-rendering;
+  @include reset;
```
Comment (VincentSmedinga) [inline packages/css/src/components/table-of-contents/table-of-contents.scss:36]: Switching these blocks completes the declarations for the list before styling another element.

```suggestion

  @include text-rendering;
  @include reset;

  .ams-table-of-contents__list {
    padding-block-start: var(--ams-table-of-contents-list-gap);
    padding-inline-start: var(--ams-table-of-contents-list-list-padding-inline-start);
  }
```

File: packages/react/src/TableOfContents/TableOfContentsLink.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+export type TableOfContentsLinkProps = {
+  label: string
```
Comment (VincentSmedinga) [inline packages/react/src/TableOfContents/TableOfContentsLink.tsx:13]:
```suggestion
  /** The text for the link. */
  label: string
```

File: storybook/src/components/TableOfContents/TableOfContents.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+export const Default: Story = {}
+
+export const MultiLevel: Story = {
+  args: {
+    title: 'Inhoudsopgave',
+    children: [
+      <TableOfContents.Link href="#section-1" key={1} label="Waarom is het belangrijk om helder te schrijven?" />,
+      <TableOfContents.Link href="#section-2" key={2} label="Schrijf in stappen">
+        <TableOfContents>
```
[the remark this answers is not in the document]
Comment (VincentSmedinga) [reply storybook/src/components/TableOfContents/TableOfContents.stories.tsx:35]: We probably need a separate `TableOfContents.List`?

File: storybook/src/components/TableOfContents/TableOfContents.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+      <TableOfContents.Link href="#section-3" key={3} label="Structuur van een heldere tekst">
+        <TableOfContents>
+          <TableOfContents.Link href="#section-3.1" label="Opmaakeisen" />
+          <TableOfContents.Link href="#section-3.2" label="Tekstlengte">
+            <TableOfContents>
+              <TableOfContents.Link href="#section-3.2.1" label="Boodschap en achtergronden" />
+            </TableOfContents>
+          </TableOfContents.Link>
+          <TableOfContents.Link href="#section-3.2" label="Alinea's en tussenkopjes" />
+          <TableOfContents.Link href="#section-3.2" label="Opsommingen" />
```
Comment (VincentSmedinga) [inline storybook/src/components/TableOfContents/TableOfContents.stories.tsx:50]:
```suggestion
          <TableOfContents.Link href="#section-3.3" label="Alinea's en tussenkopjes" />
          <TableOfContents.Link href="#section-3.4" label="Opsommingen" />
```

File: proprietary/tokens/src/components/ams/table-of-contents.tokens.json (hunk 8 above)
Comment (dlnr) [reply proprietary/tokens/src/components/ams/table-of-contents.tokens.json:8]: In the amsterdam.nl figma, guess i've taken the value of the design system figma.

**Review by VincentSmedinga — DISMISSED** [review]

File: packages/react/src/TableOfContents/TableOfContents.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (dlnr) [reply packages/react/src/TableOfContents/TableOfContents.tsx]: Changed to `ul`.

File: storybook/src/components/TableOfContents/TableOfContents.stories.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (dlnr) [reply storybook/src/components/TableOfContents/TableOfContents.stories.tsx]: Added a `List` component.

File: proprietary/tokens/src/components/ams/table-of-contents.tokens.json (hunk 2 above)
Comment (dlnr) [reply proprietary/tokens/src/components/ams/table-of-contents.tokens.json]: Checked

**Review by VincentSmedinga — APPROVED** [review]

**Pull request merged**